rust/main/chains/hyperlane-ton/src/client/provider.rs

```diff
@@ -0,0 +1,773 @@
+    },
+    utils::conversion::ConversionUtils,
+};
+use tonlib_core::TonAddress;
+use url::Url;
+
+#[derive(Clone, new)]
+pub struct TonProvider {
+    pub http_client: Client,
+    pub connection_conf: TonConnectionConf,
+    pub domain: HyperlaneDomain,
+}
+
+impl TonProvider {
+    async fn post_request(
+        &self,
+        url: Url,
+        params: &Value,
+    ) -> Result<Response, ChainCommunicationError> {
+        self.http_client
+            .post(url)
+            .header("accept", "application/json")
+            .header("Content-Type", "application/json")
+            .header("X-API-Key", self.connection_conf.api_key.clone())
+            .json(params)
+            .send()
+            .await
+            .map_err(|e| {
+                warn!("Error sending request: {:?}", e);
+                ChainCommunicationError::Other(HyperlaneCustomErrorWrapper::new(Box::new(e)))
+            })
+    }
+
+    async fn query_request<T: serde::Serialize + ?Sized>(
+        &self,
+        url: Url,
+        params: &T,
+    ) -> Result<Response, reqwest::Error> {
+        self.http_client
+            .get(url)
+            .query(params)
+            .header("accept", "application/json")
+            .header("Content-Type", "application/json")
+            .header("X-API-Key", self.connection_conf.api_key.clone())
+            .send()
+            .await
+    }
+}
+impl std::fmt::Debug for TonProvider {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TonProvider")
+            .field("client", &self.http_client)
+            .field("domain", &self.domain)
+            .finish()
+    }
+}
+impl HyperlaneChain for TonProvider {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.clone())
+    }
+}
+
+#[async_trait]
+impl HyperlaneProvider for TonProvider {
+    async fn get_block_by_height(&self, _height: u64) -> ChainResult<BlockInfo> {
+        let response = self
+            .get_blocks(
+                -1, // Masterchain
+                None,
+                Some(_height),
+                None,
+                None,
+                None,
+                None,
+                None,
+                None,
+                None,
+                None,
+            )
+            .await
+            .map_err(|e| {
+                log::error!("Error fetching block by height {}: {:?}", _height, e);
+                ChainCommunicationError::CustomError(format!(
+                    "Failed to get block by height {}: {:?}",
+                    _height, e
+                ))
+            })?;
+
+        let block = response.blocks.first().ok_or_else(|| {
+            tracing::warn!("No blocks found in the response: {:?}", response);
+            ChainCommunicationError::CustomError("No blocks found in the response".to_string())
+        })?;
```

Comment:
We have quite a generic pattern on how define chain-specific errors and convert it into `ChainCommunicationError`. Could you please use it throughout the source code?

See example here: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/main/chains/hyperlane-sealevel/src/error.rs

Reply:
Done